stravalib/model.py

```diff
@@ -59,26 +59,49 @@
 from stravalib.strava_model import (
     Split,
     SportType,
+    SummaryClub,
     SummaryPRSegmentEffort,
     SummarySegmentEffort,
     TimedZoneRange,
 )
 
+if TYPE_CHECKING:
+    from stravalib.client import BatchedResultsIterator
+
 LOGGER = logging.getLogger(__name__)
 
 
-def lazy_property(fn):
+# TODO: stravalib/model.py:71: error: Please use "Callable[[<parameters>], <return type>]"  [misc]
+def lazy_property(fn: Callable[Any]) -> property:
     """
     Should be used to decorate the functions that return a lazily loaded
     entity (collection), e.g., the members of a club.
 
+    Parameters
+    ----------
+    fn : Callable
+        The input function that returns the lazily loaded entity (collection).
+
+    Returns
+    -------
+    property
+        A lazily loaded property.
+
+    Raises
+    ------
+    exc.UnboundEntity
+        If the object is unbound (self.bound_client is None) or has a None ID (self.id is None).
+
+    Notes
+    -----
     Assumes that fn (like a regular getter property) has as single
     argument a reference to self, and uses one of the (bound) client
     methods to retrieve an entity (collection) by self.id.
+
     """
 
     @wraps(fn)
-    def wrapper(obj):
+    def wrapper(obj: Any) -> Any:
```

Comment:
`Any` works but i'm not sure if this is explicit enough here.

Reply:
This decorator is quite complex as it turns the function into a property. I've tried to come up with a compromise between  precision and not bloating the code too much.

The best solution would be to be able to say that lazy property returns `property[T | None` but property is not generic so we can't to that.

Instead I'm pretending the return type is simply a `T | None` and use an ignore comment. I think it should do the job.

With this fix I go down to 15 errors! Let me know if you have any questions.
```diff
diff --cc stravalib/model.py
index b23d0c2,6df6e62..0000000
--- a/stravalib/model.py
+++ b/stravalib/model.py
@@@ -22,6 -22,6 +22,7 @@@ from typing import 
      ClassVar,
      Literal,
      Optional,
++    cast,
      Tuple,
      Union,
      get_args,
@@@ -69,8 -75,11 +76,10 @@@ if TYPE_CHECKING
  LOGGER = logging.getLogger(__name__)
  
  
- # TODO: how do you properly type this with more precision than Any?
- def lazy_property(fn: Callable[Any], Any) -> property:
+ T = TypeVar("T")
+ U = TypeVar("U", bound="BoundClientEntity")
+ 
 -# TODO: stravalib/model.py:71: error: Please use "Callable[[<parameters>], <return type>]"  [misc]
 -def lazy_property(fn: Callable[[U], T]) -> Callable[[U], T]:
++def lazy_property(fn: Callable[[U], T]) -> T | None:
      """
      Should be used to decorate the functions that return a lazily loaded
      entity (collection), e.g., the members of a club.
@@@ -99,13 -108,13 +108,13 @@@
      """
  
      @wraps(fn)
-     def wrapper(obj: Any) -> Any:
 -    def wrapper(obj: U) -> T:
++    def wrapper(obj: U) -> T | None:
          try:
              if obj.bound_client is None:
                  raise exc.UnboundEntity(
                      f"Unable to fetch objects for unbound {obj.__class__} entity."
                  )
--            if obj.id is None:
++            if obj.id is None: # type: ignore[attr-defined]
                  LOGGER.warning(
                      f"Cannot retrieve {obj.__class__}.{fn.__name__}, self.id is None"
                  )
@@@ -116,7 -125,7 +125,7 @@@
                  f"Unable to fetch objects for unbound {obj.__class__} entity: {e}"
              )
  
--    return property(wrapper)
++    return property(wrapper) # type: ignore[return-value]
  
  
  # Custom validators for some edge cases:

```

Reply:
ok this one is really complex. so first you define type var's U and T. And U has a bound object associated with it - the BoundClient object. 
```
T = TypeVar("T") 
U = TypeVar("U", bound="BoundClientEntity")
```

then for the lazy_property definition you define the input as a Callable that is passed U and returns T. 

```
def lazy_property(fn: Callable[[U], T]) -> T | None:
```

i have two questions

1. Question: why wouldn't  `fn` return `T | None` as well?
2. if we have the T or None return why do you need this:

`type: ignore[return-value]`

Reply:
1. We don't know anything about the function `fn` that would be passed into the decorator so we can't assume that it could return `None`. In the wrapper there is a possibility to return `None` so we made the return type a union of `T` and `None`
2. If you try to remove the comment it wil complain about returning a `property` not a `T | None`. We wan't the return type to be `T | None` and not property since it will be more useful later. See more detail here https://stackoverflow.com/questions/75281747/how-to-annotate-a-decorator-that-returns-a-property
in particular, the second answer